File: aepsych/networkDisk_tests/recipient.py
"""
Created on Tue Jan  7 23:45:21 2025

@author: brainardlab-adm
"""
# Machine-specific paths are read from local_config.json at the aepsych/ root.
# That file is gitignored. Copy local_config.json.template to local_config.json
# and fill in paths for your machine before running.

import jax

jax.config.update("jax_enable_x64", True)
import os  # noqa: E402
import time  # noqa: E402
from pathlib import Path  # noqa: E402

from aepsych_dconfig import MachineConfig  # noqa: E402

machine = MachineConfig.from_json(Path(__file__))

import glob  # noqa: E402

from analysis.utils_communication import (  # noqa: E402
    CommunicateViaTextFile,
)

subject_id = int(input("Subject ID: "))
subject_init = input("Subject initials: ").strip()
session_today = int(input("Session number today: "))

# Wait for sender to create the session file
path_sub = os.path.join(machine.network_disk_path, f"sub{subject_id}")
pattern = os.path.join(
    path_sub, f"sub{subject_id}_{subject_init}*session{session_today}*.txt"
)

wait_timeout = 120
print(f"Waiting for session file matching {pattern} ...")
t_start = time.time()
while True:
    matches = glob.glob(pattern)
    if matches:
        break
    if time.time() - t_start > wait_timeout:
        raise TimeoutError(f"Timed out waiting for session file matching {pattern}")
    time.sleep(0.5)

file_name = os.path.basename(max(matches, key=os.path.getmtime))
print(f"Found session file: {file_name}")

# Initialize communication class
communicator = CommunicateViaTextFile(
    path_sub,
    retry_delay=3 / 60,  # 1 frame
    timeout=1200,
)  # 1200s = 20 mins
communicator.check_and_handle_file(file_name)

# Step 1: Wait for Initialization
print("Waiting for initialization command...")
communicator.confirm_communication()
print("Initialization confirmed.")

# Step 2: Wait for and confirm RGB values
trial_counter = 0
while True:
    if communicator.terminate:
        break
    print(f"Trial #{trial_counter}...")
    communicator.confirm_RGBvals(response_delay=0.1)
    trial_counter += 1
    time.sleep(0.01)
    print("RGB values confirmed.")

File: aepsych/generic/networkdisktest/recipient.m
% recipient.m
% MATLAB test stand-in for recipient.py.
%
% Reads network_disk_path from local_config.json in the same directory.
% Copy local_config.json.template -> local_config.json and fill in the path.
%
% Responses are random (0 or 1).

%% Add communication class to path
here = fileparts(mfilename('fullpath'));
addpath(fullfile(here, '..', '..', 'matlab'));

%% Read config
cfg = jsondecode(fileread(fullfile(here, 'local_config.json')));
network_disk_path = cfg.network_disk_path;

%% Get experiment info
dlg_answer = inputdlg( ...
    {'Subject ID:', 'Subject Initials:', 'Session number today:'}, ...
    'Experiment Info', 1, {'', '', ''});
if isempty(dlg_answer)
    error('No experiment info entered.');
end
subject_id    = str2double(dlg_answer{1});
subject_init  = strtrim(dlg_answer{2});
session_today = str2double(dlg_answer{3});

%% Wait for sender to create the session file
path_sub = fullfile(network_disk_path, sprintf('sub%d', subject_id));
pat      = sprintf('sub%d_%s*session%d*.txt', subject_id, subject_init, session_today);

full_file_path = WPPMCommunicator.waitForSessionFile(path_sub, pat);

%% Run communication
comm = WPPMCommunicator(full_file_path);

fprintf('Waiting for initialization command...\n');
comm.confirmCommunication();
fprintf('Initialization confirmed.\n');

trial_counter = 0;
while ~comm.terminate
    comm.confirmRGBvals(0.1);
    trial_counter = trial_counter + 1;
    fprintf('Trial #%d confirmed.\n', trial_counter);
end

fprintf('Communication complete (%d trials).\n', trial_counter);
